clients/csharp-uwp/UWPVoiceAssistantSample/KeywordRegistration.cs

```diff
@@ -1,278 +1,231 @@
-        /// <returns>A <see cref="Task"/> that returns on successful keyword setup.</returns>
-        public async Task<ActivationSignalDetectionConfiguration> UpdateKeyword(
-            string keywordDisplayName,
-            string keywordId,
-            string keywordModelId,
-            string keywordActivationModelDataFormat,
-            string keywordActivationModelFilePath,
-            Version availableActivationKeywordModelVersion,
-            string confirmationKeywordModelPath)
-        {
-            this.KeywordDisplayName = keywordDisplayName;
-            this.KeywordId = keywordId;
-            this.KeywordModelId = keywordModelId;
-            this.KeywordActivationModelDataFormat = keywordActivationModelDataFormat;
-            this.KeywordActivationModelFilePath = keywordActivationModelFilePath;
-            this.AvailableActivationKeywordModelVersion = availableActivationKeywordModelVersion;
-            this.ConfirmationKeywordModelPath = confirmationKeywordModelPath;
-
-            this.keywordConfiguration = null;
-            return await this.GetOrCreateKeywordConfigurationAsync();
-        }
-
-        /// <summary>
-        /// Fetches and, if necessary, creates an activation keyword configuration matching the
-        /// specified keyword registration information.
-        /// </summary>
-        /// <returns>A <see cref="Task"/> that returns on successful keyword setup.</returns>
-        public async Task<ActivationSignalDetectionConfiguration> GetOrCreateKeywordConfigurationAsync()
-        {
-            using (await this.creatingKeywordConfigSemaphore.AutoReleaseWaitAsync())
-            {
-                if (this.keywordConfiguration != null)
-                {
-                    return this.keywordConfiguration;
-                }
+    using Windows.ApplicationModel.ConversationalAgent;
+    using Windows.Storage;
+
+    /// <summary>
+    /// A helper class to encapsulate the process for registering a activation keyword.
+    /// </summary>
+    public class KeywordRegistration : IKeywordRegistration, IDisposable
+    {
+        private readonly SemaphoreSlim creatingKeywordConfigSemaphore;
+        private ActivationSignalDetectionConfiguration softwareKeywordConfiguration;
+        private ActivationSignalDetectionConfiguration hardwareKeywordConfiguration;
+        private bool disposed;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="KeywordRegistration"/> class.
+        /// </summary>
+        /// <param name="availableActivationKeywordModelVersion">Version of the most recent keyword model that is available.</param>
+        public KeywordRegistration(Version availableActivationKeywordModelVersion)
+        {
+            this.AvailableActivationKeywordModelVersion = availableActivationKeywordModelVersion;
+
+            this.creatingKeywordConfigSemaphore = new SemaphoreSlim(1, 1);
+
+            _ = this.InitializeConfigurations();

```

Comment:
Minor: given this returns a `Task`, would `InitializeConfigurationAsync` be appropriate?